proposed/simple-csharp-programs-motivation-walkthrough.md

```diff
@@ -0,0 +1,311 @@
+# Simple C# Programs
+
+This speclet defines "Simple C# Programs" and covers what they can do from the user's standpoint. It does **not** get into detailed design points, a deep dive of a particular mechanisms, etc. The goal is to drive consensus on the user experience and then walk back towards the right design and implementation.
+
+It is also designed to be incremental. Not everything written here must be implemented. For example, we could ship a version without support for packages/frameworks.
+
+## Motivation
+
+C# has lots of great things going for it, but it falls short in a key area: getting started with the least possible friction.
+
+Lots of developers in the world have experience with languages like JavaScript, Python, and Ruby. In these languages, you typically get started in one of these three ways:
+
+1. Launch a REPL and start typing code and getting results
+2. Create a file with the right extension and run the code in that file
+3. Use build tooling to create an application and then run that application
+
+If you look at online guides for the above languages, they tend to start people off with (1) and (2) and rarely with (3).
+
+In the C# world, we start people off with (3) with the base tooling, the .NET CLI.
+
+This is problematic, because these people **do not expect that they need to do (3) to get started**. The tools don't support their instincts and there is very little information online to support them.
+
+Of course, REPLs are another great way to get started. But they are outside the scope of this document.
+
+## What is a Simple C# Program
+
+A Simple C# Program is this:
+
+* A C# file
+* A set of global `using`s inferred by the build environment
+* A tool that can run that file / produce a working program
+* A way to incorporate libraries and frameworks/sdks into that file and run it
+* A way to "graduate" to using a project file
+
+A Simple C# Program is **not** a replacement for project-based C# code or .NET Interactive.
+
+## Simple C# is just C# code
+
+Simple C# Programs are not a dialect of the C# language, like the interactive dialect, CSX. Any code in a Simple C# Program can be built and run in a project-based program.
+
+## User experience walkthrough
+
+Given the following code in `Hello.cs`:
+
+```csharp
+WriteLine("Hello World");
+```
+
+You can just run it:
+
+```bash
+$ dotnet run
+Hello World
+```
+
+Or with a shebang,
+
+```csharp
+#!/usr/bin/shared/dotnet
+WriteLine("Hello World");
+```
+
+and running:
+
+```bash
+$ ./hello.cs
+Hello World
+```
+
+The `using`s that are normally in a C# project template are implicitly available at design-time. This is accomplished with [Global Using Directives](https://github.com/dotnet/csharplang/blob/main/proposals/GlobalUsingDirective.md) that the build tooling will have available.
+
+The shebang is covered in [Shebang (#!) support](https://github.com/dotnet/csharplang/issues/3507).
```

Comment:
The term shebang originates in the US Civil War when it was coined by the Confederate soldiers [with the meaning of](https://en.wiktionary.org/wiki/shebang) "temporary shelter".

[Wikipedia](https://en.wikipedia.org/wiki/Shebang_(Unix)) provides several other names for `#!` that might be more inclusive: sha-bang, hashbang, pound-bang, or hash-pling.

Reply:
TIL...

Reply:
FWIW Wikipedia also says the shebang name might have been coined for Unix scripts independently of that earlier usage, as a contraction of hash bang or sharp bang.

Reply:
The name can easily change, although we'll have to balance this out with following/not following a convention used in UNIX land. For now, since the C# proposal uses the name, I'll leave it as-is. If we decide to change the name for other reasons, it will change here.

Reply:
> convention used in UNIX land

The Wikipedia article mentions several other names for this that are already established  - we can just pick one of those names. A soldier's "temporary shelter" in a war is often someone else's "home", and they often don't wait for an invitation. I see no good reason to use that language in our repos if there's established alternatives.

It should be easy enough to change in the C# spec as well - @333fred what do you think?

Reply:
It's changed.

Reply:
On this end too